File: runSuite.py
import argparse
import os

def fileExist(filename):
    try:
        with open(filename, 'r') as f:
            return True
    except IOError as e:
        return False

def runTest(filename, program, directory, noError):
    input = directory + filename + ".in"
    args = directory + filename + ".args"
    output = directory + filename + ".out"
    error = " 2> /dev/null" if noError else ""

    if not fileExist(output):
        print("Could not find file", output + ". Omitting test", filename + ".")
        return

    actual = os.popen("mktemp").read().strip()

    if fileExist(input) and fileExist(args):
        os.system(program + " $(cat " + args + ") < " + input + " > " + actual + error)
    elif fileExist(input):
        os.system(program + " < " + input + " > " + actual + error)
    elif fileExist(output):
        os.system(program + " $(cat " + args + ")  > " + actual + error)
    else:
        os.system(program + " > " + actual + error)
    
    if not open(actual).read() == open(output).read():
        print("Test failed:", filename)
        print("Args:")
        if fileExist(args):
            os.system("cat " + args)
        print("Input:")
        if fileExist(input):
            os.system("cat " + input)
        print("Expected:")
        os.system("cat " + output)
        print("Actual:")
        os.system("cat " + actual)

    os.system("rm " + actual)

def main():
    path = os.popen("pwd").read().strip()

    parser = argparse.ArgumentParser(description="Runs input/output tests for a given program.")
    parser.add_argument('program', help="Name of the program to be tested")
    parser.add_argument('-f', '--suitefile', nargs='?', default="suite.txt", help="Text file with name stubs of test files. Defaults to suite.txt")
    parser.add_argument('-d', '--directory', nargs='?', default=path, help="Path to directory with suitefile and test files from current directory. Defaults to current directory.")
    parser.add_argument('-e', '--error', action="store_true", help="Surpress output to stderr")

    args = vars(parser.parse_args())

    suiteFile = args['suitefile']
    programName = args['program']
    directory = args['directory']
    error = args['error']
    
    if directory[-1:] != '/':
        directory += "/"

    suiteFile = directory + suiteFile

    if fileExist(suiteFile):
        for filename in open(suiteFile).readlines():
            runTest(filename.strip(), programName, directory, error)
    else:
        print("Could not open " + suiteFile)
        return

main()

File: runSuite.cc
#include <iostream>
#include <sys/wait.h>
#include <fstream>   
#include <string>
#include <cstdlib>
#include <cstdio>

int system_wrapper(std::string cmd) {
        int ret = system(cmd.c_str());
        return WEXITSTATUS(ret);
}

// tests to see if a file exists in the current directory
bool file_test(const std::string &filename) {
	return system_wrapper("test -f " + filename) == 0;
}

// gets the next filename from f
bool get_filename(std::ifstream &f, std::string &name) {
	f >> std::noskipws;
	name = "";
	char c;

	while (f >> c) {
		if (c == ' ' || c == '\n') {
			return true;
		} else {				
			name += c;
		}
	}
	return false;
}

// determines if name is a real filename or just whitespace
bool real_filename(std::string &name) {
	for (unsigned int i = 0; i < name.size(); ++i) {
		if (name.at(i) != ' ' && name.at(i)!= '\n') {
			return true;
		}
	}
	return false;
}	

// runs the file test on program
bool run_test(std::string &filename, std::string &program) {
	bool args_exist = file_test(filename + ".args");
	bool in_exist = file_test(filename + ".in");
	
	const std::string args = filename + ".args";
	const std::string input = filename + ".in";
	const std::string output = filename + ".out";

	if (!file_test(output)) {
		std::cerr << output + " does not exist or is not readable" << std::endl;
		return false;
	}
	
	#if 0
	std::ifstream args_file{args};
	std::ifstream input_file{input};

	args_file >> std::noskipws;
	
	char m ='\0';
	while (args_file >> m) {
		if (m != ' ' && m != '	' && m != '\n') {
			args_exist = true;
			break;
		}
		args_exist = false;
	}
	if (m == '\0') args_exist = false;
	#endif

	FILE *f1 = popen("mktemp", "r");
	char a[200];
	fscanf(f1, "\%s", a);
	pclose(f1);
	std::string actual = std::string(a);

	if (args_exist && in_exist) {
		system_wrapper(program + " $(cat " + args + ") < " + input + " > " + actual);
	} else if (args_exist) {
		system_wrapper(program + " $(cat " + args + ") > " + actual);
	} else if (in_exist) {
		system_wrapper(program + " < " + input + " > " + actual);
	} else {
		//system_wrapper(program + " > " + actual);
		std::cerr << "Please provide an input or arguments file" << std::endl;
		return 5;
	}

	std::ifstream x{actual};
	std::ifstream y{output};
	char c;
	char d;
	bool ret = false;
	while (x >> c && y >> d) {
		if (c != d) {
			ret = true;
			break;
		}
	}
	y >> d;
	ret = c != d;

	if (ret) {
		std::cout << "Test failed: " << filename << std::endl;
		std::cout << "Args:" << std::endl;
		if (args_exist) system_wrapper("cat " + args);
		std::cout << "Input:" << std::endl;
		if (in_exist) system_wrapper("cat " + input);
		std::cout << "Expected:" << std::endl;
		system_wrapper("cat " + output);
		std::cout << "Actual:" << std::endl;
		system_wrapper("cat " + actual);
	}

	system_wrapper("rm " + actual);
	return true;
}

int main(int argc, char **argv) {
	if (argc != 3) {
		std::cerr << "Incorrect number of arguments supplied to runSuite" << std::endl;
		return 1;
	} else if (!file_test(argv[1])) {
		std::cerr << std::string(argv[1]) + " does not exist or is not readable" << std::endl;
		return 2;
	}

	std::ifstream suitefile{argv[1]};
	std::string filename;
	std::string program = std::string(argv[2]);

	while (get_filename(suitefile, filename)) {
	       if (real_filename(filename)) {
		       if (!run_test(filename, program)) return 3;
	       }
	}
	return 0;
}
